### ISSUE TITLE
[SDL3] GPU: Investigate adding a Query API

### ISSUE THREAD

**Issue by WoozChucky** [libsdl-org/SDL#11696]
Not sure if other supported apis have the same concept of Vulkan's `VkQueryPool`s, but would be great to have this exposed (via some SDL api wrapping) somehow to be able to query some gpu information.

**Comment by thatcosmonaut** [comment]
What's your use case? We would have to do some background research to identify a good cross-platform API surface for this.

**Comment by WoozChucky** [comment]
Mostly timing measurements in the rendering pipelines.

```c++
VK_CHECK(vkGetQueryPoolResults(device, queryPool, 0, ARRAYSIZE(queryResults), sizeof(queryResults), queryResults, sizeof(queryResults[0]), VK_QUERY_RESULT_64_BIT));

double frameGpuBegin = double(queryResults[0]) * props.limits.timestampPeriod * 1e-6;
double frameGpuEnd = double(queryResults[1]) * props.limits.timestampPeriod * 1e-6;
```

Once again, this might be too specific for Vulkan and troublesome/inexistent in other apis (I havent messed with Metal or DX at all).

**Comment by thatcosmonaut** [comment]
This is maybe doable but Metal only supports very coarse timing queries. I'll flag this for 3.x in case anyone wants to investigate this after 3.2 is released.

**Comment by savant117** [comment]
I've been thinking about this more, and put together an API proposal for this. Looking for feedback before I start writing any code.

I propose adding a new queries type which supports timestamps, occlusion, and binary occlusion. Pipeline statistics are not included because support varies too much. Support for the proposed query types is very high, and I think any device which already supports SDL GPU will support these queries.

Timestamp queries: https://vulkan.gpuinfo.org/displaydevicelimit.php?platform=android&name=timestampPeriod
Occlusion queries: https://vulkan.gpuinfo.org/listfeaturescore10.php

There were some questions about metal support for timestamps, but I believe using this API will give precise results (this is what MoltenVK now uses also): https://developer.apple.com/documentation/metal/sampling-gpu-data-into-counter-sample-buffers

Occlusion queries have a begin / end pair, with draw calls placed between them. Binary queries simply return 1 when ANY pixel is visible for the measured draw calls, where as the non binary (precise) query counts the number of visible pixels. For timestamps, either begin or end can be called, as it simply records the timestamp. For timestamps, there is an additional function which gets the GPU timestamp frequency. This is required to convert the timestamps into something usable. 

There are two functions to retrieve the results. One function results results to the CPU, which can optionally wait for the results to be ready. The other function directly copies the results to a gpu buffer, which can then be used for indirect rendering or as input to compute.

There is no emulation required in the backends and it will be a very thin wrapper. The most complex detail is that for DX12 backend, SDL_GetGPUQueryResults will have to use a fence and staging buffer internally to keep track of result readiness whereas vulkan and metal have APIs that do this automatically. 

Here is the proposed API:
`
typedef struct SDL_GPUQueries SDL_GPUQueries;

typedef enum {
    SDL_GPU_QUERY_TIMESTAMP,
    SDL_GPU_QUERY_OCCLUSION,
    SDL_GPU_QUERY_BINARY_OCCLUSION
} SDL_GPU_QueryType;

/* Create a query pool or some type with some number of available queries */
SDL_GPUQueries* SDL_CreateGPUQueries(SDL_GPUDevice *device, SDL_GPU_QueryType type, UInt32 queryCount);

/* Destroy a query pool */
void SDL_DestroyGPUQueries(SDL_GPUQueries* queries);

/* Begin an occlusion query at a specific index, or records a timestamp  */
void SDL_BeginGPUQuery(SDL_GPUCommandBuffer *command_buffer, SDL_GPUQueries* queries, UInt32 queryIndex);

/* Ends an occlusion query at a specific index, or records a timestamp */
void SDL_EndGPUQuery(SDL_GPUCommandBuffer *command_buffer, SDL_GPUQueries* queries, UInt32 queryIndex);

/* Retrieve query results. Put results (uint64 or uint32 for timestamp or occlusion respectively) into results. Optionally block and wait for results to finish. Returns true if results were written, or false if results were not ready */
Bool SDL_GetGPUQueryResults(SDL_GPUQueries* queries, UInt32 firstQuery, UInt32 count, void* results,  Bool wait);

/* Copy query results directly to a GPU buffer */
void SDL_CopyGPUQueryResultsToBuffer(SDL_GPUCopyPass* copy_pass, SDL_GPUQueries* queries, UInt32 firstQuery, UInt32 count, SDL_GPUBuffer* destBuffer, UInt32 bufferOffset);

/* Get GPU timestamp frequency, needed to compute actual wall clock times from timestamps */
UInt64 SDL_GetGPUTimestampFrequency(SDL_GPUDevice* device);
`

**Comment by thatcosmonaut** [comment]
This looks mostly right to me. I have a few notes:

- Rename SDL_GPUQueries to SDL_GPUQueryPool.
- I don't like having two ways to do the same thing and the API should mirror how data retrieval works for other functions. I think that the only way should be to call the copy function and DownloadFromBuffer and then you can use a fence to know when it's ready.

**Comment by savant117** [comment]
Ok makes sense. Here is the updated proposal:

```
typedef struct SDL_GPUQueryPool SDL_GPUQueryPool;

typedef enum {
    SDL_GPU_QUERY_TIMESTAMP,
    SDL_GPU_QUERY_OCCLUSION,
    SDL_GPU_QUERY_BINARY_OCCLUSION
} SDL_GPU_QueryType;

/* Create a query pool of some query type with some number of available queries */
SDL_GPUQueryPool* SDL_CreateGPUQueryPool(SDL_GPUDevice *device, SDL_GPU_QueryType type, UInt32 query_count);

/* Destroy a query pool */
void SDL_ReleaseGPUQueryPool(SDL_GPUQueryPool* query_pool);

/* Begin an occlusion query at a specific index, or records a timestamp  */
void SDL_BeginGPUQuery(SDL_GPUCommandBuffer *command_buffer, SDL_GPUQueryPool* query_pool, UInt32 query_index);

/* Ends an occlusion query at a specific index, or records a timestamp */
void SDL_EndGPUQuery(SDL_GPUCommandBuffer *command_buffer, SDL_GPUQueryPool* query_pool, UInt32 query_index);

/* Copy query results directly to a GPU buffer */
void SDL_CopyGPUQueryResultsToBuffer(SDL_GPUCopyPass* copy_pass, SDL_GPUQueryPool* query_pool, UInt32 first_query, UInt32 count, SDL_GPUBuffer* dest, UInt32 dest_offset);

/* Get GPU timestamp frequency, needed to compute actual wall clock times from timestamps */
UInt64 SDL_GetGPUTimestampFrequency(SDL_GPUDevice* device);
```

**Comment by thatcosmonaut** [comment]
Looks good.

**Comment by savant117** [comment]
Made a fork to start working on this. Will make a PR once it's a little further along. Can follow the progress here: https://github.com/savant117/SDL/tree/queries

Got the basics set up, and will start on the DX12 backed next. If anyone wants to help speed it along, feel free to contribute! Could maybe use some help with the metal backend, since I don't have a mac.

**Comment by slouken** [comment]
@savant117, this looks good, how's it going?

**Comment by mr1name** [comment]
Today I was tinkering with counter sample buffers in Metal and threw together a very simple and dirty [implementation](https://github.com/libsdl-org/SDL/compare/main...mr1name:SDL:metal-counter-samples) for my debugging needs. Of course I’m not proposing the resulting “API” or the implementation itself, since I didn’t even try to align with Vulkan or DX, but some observations might still be useful.
https://github.com/libsdl-org/SDL/compare/main...mr1name:SDL:metal-counter-samples

Even though Metal’s API formally documents a broad set of boundaries at which you can sample counters, if you run a test query on any Mac with an M-series chip, you’ll find that only atStageBoundary is actually supported.

<img width="933" height="111" alt="Image" src="https://github.com/user-attachments/assets/91c58654-6278-4339-8238-87b7ed56f8f0" />

https://developer.apple.com/documentation/metal/sampling-gpu-data-into-counter-sample-buffers

This is probably important to keep in mind, because it can potentially impose constraints on a generalized API. To sample via atStageBoundary, you have to specify the output buffer indices at compute/render pass initialization time.

e.g.
For a render pass:
```objective-c
MTLRenderPassSampleBufferAttachmentDescriptor *attr = passDescriptor.sampleBufferAttachments[0];
attr.sampleBuffer = query.buffer->handle;
attr.startOfVertexSampleIndex   = query.startTimeIndex;
attr.endOfVertexSampleIndex     = MTLCounterDontSample;
attr.startOfFragmentSampleIndex = MTLCounterDontSample;
attr.endOfFragmentSampleIndex   = query.endTimeIndex;
...
metalCommandBuffer->renderEncoder = [metalCommandBuffer->handle renderCommandEncoderWithDescriptor:passDescriptor];
```
e.g.
For a compute pass:
```objective-c
MTLComputePassDescriptor *cpDesc = [MTLComputePassDescriptor computePassDescriptor];
            MTLComputePassSampleBufferAttachmentDescriptor *attr = cpDesc.sampleBufferAttachments[0];
attr.sampleBuffer = query.buffer->handle;
attr.startOfEncoderSampleIndex = query.startTimeIndex;
attr.endOfEncoderSampleIndex = query.endTimeIndex;
...
metalCommandBuffer->computeEncoder = [metalCommandBuffer->handle computeCommandEncoderWithDescriptor:cpDesc]; 
```

So, for example, an API design with separate Begin/End calls, where the timestamp from End could be redirected into a different buffer slot, likely can’t be represented precisely in Metal. Unless you treat the Begin/End range as the sum of samples written in between via atStageBoundary, and specify the output pair already at Begin; or at the driver level immediately record (end_ts - begin_ts) and expose only a single value to the user; or create some extra intermediate buffer.

In any case, it would still be nice to have at least some kind of API for these purposes, even if it ends up being quite limited in flexibility compared to any of the backends.

**Comment by DomGries** [comment]
This feature would be great. GPU queries are the only way to get actual GPU-side timing, occlusion results and pipeline statistics. All three current backends support this (Vulkan `VkQueryPool`, D3D12 `ID3D12QueryHeap`, Metal `MTLCounterSampleBuffer`). Without it, anyone needing profiling, occlusion culling, adaptive quality or frame pacing must bypass SDL entirely and call the native API directly, breaking portability.

**Comment by raddad772** [comment]
I'd love it too, working on Dreamcast emulator and occlusion queries would make non-compute OIT at lot easier.

**Comment by jdolan** [comment]
Pasting the closed/duplicate #15929 I created for this today, because I think there's some useful additional input around how the API should behave and expected usage to avoid pipeline stalls. I think these things should be core to the implementation, and well documented so that users understand how to use them. 

> I'm in the process of porting [Quetoo](https://github.com/jdolan/quetoo) to SDL_gpu. The port is going well, but I’ve hit a performance bottleneck due to the lack of hardware occlusion query support.
> 
> I'm aware of more modern alternatives like Hierarchical-Z with indirect rendering. But that's a complete rewrite of my renderer, not a port. Given SDL_gpu's [self-declared target of roughly 10 year old feature set](https://github.com/libsdl-org/SDL/issues/12094#issuecomment-2893116922), occlusion queries seem sorely absent. All three primary backends already support this functionality natively:
> 
> * Vulkan: [VkQueryPool](https://docs.vulkan.org/refpages/latest/refpages/source/VkQueryPool.html)
> * Metal: [MTLVisibilityResultBuffer](https://developer.apple.com/documentation/metal/mtlrenderpassdescriptor/visibilityresultbuffer)
> * Direct3D 12: [ID3D12QueryHeap](https://microsoft.github.io/DirectX-Specs/d3d/CountersAndQueries.html) (with D3D12_QUERY_TYPE_OCCLUSION)
> 
> **Proposed Approach: _Deferred Retrieval_**
> 
> To keep this implementation "zero-stall" and consistent with SDL_gpu’s current design, I propose a "deferred retrieval" model. In this model, the application issues a query in frame N, and retrieves the result during frame N+1. This pattern effectively hides GPU-to-CPU latency, avoiding the need for explicit fences or pipeline stalls.
> 
> This approach mimics the standard workflow found in traditional OpenGL implementations and is well-supported by the underlying hardware APIs.
> 
> **Potential API Surface:**
> ```c
> /**
>  * @brief Creates a sized @c SDL_GPUQueryPool.
>  * @param device The @c SDL_GPUDevice.
>  * @param size The desired pool size.
>  * @returns A handle to a new @c SDL_GPUQueryPool, or NULL on failure.
>  */
> SDL_GPUQueryPool* SDL_CreateGPUQueryPool(SDL_GPUDevice *device, Uint32 size);
> 
> /**
>  * @brief Begins an @c SDL_GPUQuery.
>  * @param pool The @c SDL_GPUQueryPool.
>  * @returns A handle to the allocated @c SDL_GPUQuery, or NULL on failure.
>  * @remarks Only valid within an active @c SDL_GPURenderPass.
>  */
> SDL_GPUQuery *SDL_BeginGPUQuery(SDL_GPUQueryPool *pool);
> 
> /**
>  * @brief Ends an @c SDL_GPUQuery. 
>  * @param query The @c SDL_GPUQuery.
>  * @remarks Only valid within an active @c SDL_GPURenderPass.
>  */
> void SDL_EndGPUQuery(SDL_GPUQuery *query);
> 
> /**
>  * @brief Retrieves the results of an @c SDL_GPUQuery.
>  * @param query The @c SDL_GPUQuery.
>  * @param result A pointer to a @c Uint64 to store the sample count.
>  * @returns true if the result is available, false if it is still pending.
>  * @remarks This is non-blocking, and should be called before conditionally reissuing the query. 
>  *          Upon returning true, the query is immediately returned to the pool for further work.
>  */
> bool SDL_GetGPUQueryResult(SDL_GPUQuery *query, Uint64 *result);
> ```
> **Example Usage**
> 
> The following pattern allows for a deferred retrieval workflow, where the CPU checks for results from a previous frame to avoid pipeline stalls.
> 
> ```c
> static SDL_GPUQueryPool *pool;
> 
> void init(SDL_GPURenderDevice *device) {
>   pool = SDL_CreateGPUQueryPool(device, SDL_arraysize(my_expensive_objects));
> }
> 
> // probably the very first 3D drawing in your frame
> void update_occlusion_queries(SDL_GPURenderPass *pass) {
> 
>   draw_depth_pass(pass); // draw depth-only Z pass to give the queries something to test against
> 
>   my_expensive_object *obj = my_expensive_objects;
>   for (int i = 0; i < SDL_arraysize(my_expensive_objects); i++, obj++) {
>     
>     if (obj->query) { // frame N + 1, non-blocking check for an available result
>       Uint64 res;
>       if (SDL_GetGPUQueryResult(obj->query, &obj->query_result)) {
>         // the query result is available, read it and relinquish the handle
>         obj->query = NULL;
>       }
>     }
> 
>     if (obj->query == NULL) { // frame N
>       // the object has no query (first frame, or query result just came back)
>       // re-issue the query for the current frame
>       obj->query = SDL_BeginGPUQuery(pool);
>       if (obj->query) {
>         SDL_DrawGPUPrimitives(pass, obj->num_vertices, 1, obj->first_vertex, 0);
>         SDL_EndGPUQuery(obj->query);
>       }
>     }
>   }
> }
> ```
> This API would allow [a flow like Quetoo uses](https://github.com/jdolan/quetoo/blob/main/src/client/renderer/r_occlude.c#L176) to work beautifully in SDL_gpu.
> 
> I would be happy to help contribute this feature if the team is open to it. Thank you for your hard work on SDL_gpu!

I plan to take a crack at the Metal implementation of this, based off of @thatcosmonaut's branch, in the next week or so.

**Comment by Plixo2** [comment]
@jdolan, @savant117, any updates on this?

**Comment by jdolan** [comment]
@Plixo2 Yes, I implemented occlusion queries for Metal on top of the Vulkan support that @thatcosmonaut had already started. Both are available from the pull request I opened against his fork here:

https://github.com/thatcosmonaut/SDL/pull/247

But @thatcosmonaut wanted to reevaluate the design due to a Metal requirement I documented here:

https://github.com/thatcosmonaut/SDL/pull/247#pullrequestreview-4690228884

[EDIT] I too would love to see this move forward.

**Comment by thatcosmonaut** [comment]
Yeah I decided that the best way forward with this is to add a BeginRenderPass2 call: https://github.com/libsdl-org/SDL/issues/15994

Sorry this is moving slowly but it will absolutely be in by 3.6.0 at the latest.

**Comment by jdolan** [comment]
It's all good. Can I help? Do you want me to rebase my PR against a different branch? Or should I implement timer query support on Metal to round out the feature?
